velox/common/memory/MemoryArbitrator.cpp

```diff
@@ -197,12 +197,15 @@ MemoryReclaimer::reclaim(MemoryPool* pool, uint64_t targetBytes, Stats& stats) {
     int64_t reservedBytes;
   };
   std::vector<Candidate> candidates;
-  candidates.reserve(pool->children_.size());
-  for (auto& entry : pool->children_) {
-    auto child = entry.second.lock();
-    if (child != nullptr) {
-      const int64_t reservedBytes = child->reservedBytes();
-      candidates.push_back(Candidate{std::move(child), reservedBytes});
+  {
+    folly::SharedMutex::ReadHolder guard{pool->poolMutex_};
+    candidates.reserve(pool->children_.size());
+    for (auto& entry : pool->children_) {
+      auto child = entry.second.lock();
+      if (child != nullptr) {
```

Comment:
When would a child be null?

Why not use pool->visitChildren?

How do you decide when to use the lock when not?

Reply:
The child could be null as the parent pool only holds a weak ptr. And the child pool tries to remove itself from the parent pool's child list on destruction so shared ptr upgrade will fail.

pool->visitChildren use lock but it can't sort the memory pool based on the usage. pool->visitChildren only visit child pool based on its position in parent pool's child list.

Reply:
Got it. Thank you for explaining.